PrestaShop/PrestaShop
# Migrate customer bought and viewed products to grids

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | see below
| Type?             | refacto
| Category?         | BO
| BC breaks?        | partial - CustomerForViewing object now returns empty array for bought and viewed products.
| Deprecations?     | no
| How to test?      | see below
| Fixed ticket?     | Fixes #15444
| Related PRs       | 
| Sponsor company   | 

### Details
- Right now, the bought and viewed product tables on customer details were manually rendered tables. I changed their implementation to grid components like I did with orders and carts in https://github.com/PrestaShop/PrestaShop/pull/31442.
- **This has following advantages:**
  -  Introduces pagination so there is not a kilometer of bought products and the page is actually usable for bigger clients,
  -  Makes the page load faster.
  - Allows the use of standard hooks so people can customize these tables.
  - Bought products - I added a column Order ID, for better clarity.
  - Viewed products - I added a column Cart ID with a link to the cart in backoffice.

### Behavior and wording change
Purchased products now include products from all orders, not just the valid ones. You now see also products from orders that are waiting for payment. Before, they contained only products from valid orders. If the product is bought, it's bought. Follows a simmilar information as in here - https://github.com/PrestaShop/PrestaShop/pull/31440.

Simmilarily, viewed products now contain only products from carts that have not been ordered. ( = not in ps_orders table.) Previously, it contained all products from carts that have not been ordered OR not in a valid order.

The name of `Viewed products` was changed to `Products in carts` to better reflect what it means.

### Screenshot
**Before - chrome did not even want to make me a full height screenshot :D**
https://user-images.githubusercontent.com/6097524/224347878-985f7b3c-5b15-4ff1-8d9f-2f0d7eb8edb9.jpg

**After**
![Snímek obrazovky 2023-03-13 142201](https://user-images.githubusercontent.com/6097524/224714218-31f8bc12-815e-4b47-a046-95e2c41d697a.jpg)

### How to test
- Open customer detail page in backoffice and check blocks with bought and viewed products.
- See that everything works the same, but limits the amount of shown products to 20, adds pagination, allows sorting and there is a bit more information. 
- Beware that `Viewed products` block is a bit misleading, what it really means is `Products in unordered carts`.
- Beware that you can see a bit different data with and without this PR. With this PR, you can see more products in `Bought products` and less in `Viewed products`.
- Beware that if customer has no bought products and no products in cart, the block is not shown, like before.
